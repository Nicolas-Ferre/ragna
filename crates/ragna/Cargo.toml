[package]
name = "ragna"
description = "Library for easily creating GPU-native applications"
readme = "../../README.md"
keywords = ["gpu", "rendering", "compute", "shader"]
categories = ["game-development", "graphics", "rendering", "concurrency"]
exclude = [".github", "README.md"]
version.workspace = true
authors.workspace = true
edition.workspace = true
license.workspace = true
repository.workspace = true
rust-version.workspace = true

[dependencies]
bitfield-struct.workspace = true
derive-where.workspace = true
futures.workspace = true
fxhash.workspace = true
itertools.workspace = true
once_cell.workspace = true
wgpu.workspace = true
ragna_derive.workspace = true

[lints]
workspace = true
