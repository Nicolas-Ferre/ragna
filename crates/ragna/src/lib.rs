//! A library for creating easily a GPU-native application.

mod app;
mod operations;
mod operators;
mod runner;
mod types;
mod wgsl;

pub use app::*;
pub use operators::*;
pub use types::*;

/// Transforms a Rust module to a GPU module.
pub use ragna_derive::gpu;

// TODO: adapt compilation tests
