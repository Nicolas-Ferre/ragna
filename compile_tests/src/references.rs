fn main() {}

#[ragna::gpu]
mod gpu {
    use ragna::I32;

    fn func(param: &mut I32) -> &mut I32 {
        &mut 0
    }

    fn conditional_return(value: &I32) -> &I32 {
        if true {
            value
        } else {
            value
        }
    }
}
