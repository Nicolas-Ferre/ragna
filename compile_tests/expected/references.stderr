error: `mut` keyword must not be used for references
 --> compile_tests/src/references.rs:7:21
  |
7 |     fn func(param: &mut I32) -> &mut I32 {
  |                     ^^^

error: `mut` keyword must not be used for references
 --> compile_tests/src/references.rs:7:34
  |
7 |     fn func(param: &mut I32) -> &mut I32 {
  |                                  ^^^

error: `mut` keyword must not be used for references
 --> compile_tests/src/references.rs:8:10
  |
8 |         &mut 0
  |          ^^^

error[E0277]: the trait bound `&I32: Gpu` is not satisfied
  --> compile_tests/src/references.rs:12:9
   |
12 |         if true {
   |         ^^ the trait `Gpu` is not implemented for `&I32`
   |
help: consider borrowing here
   |
12 |         &if true {
   |         +




error[E0515]: cannot return reference to temporary value
 --> compile_tests/src/references.rs:8:9
  |
8 |         &mut 0
  |         ^^^^^-
  |         |    |
  |         |    temporary value created here
  |         returns a reference to data owned by the current function